flink-core/src/main/java/org/apache/flink/util/AbstractCloseableRegistry.java

```diff
@@ -163,9 +163,9 @@ protected final void addCloseableInternal(Closeable closeable, T metaData) {
 	/**
 	 * Removes a mapping from the registry map, respecting locking.
 	 */
-	protected final void removeCloseableInternal(Closeable closeable) {
+	protected final boolean removeCloseableInternal(Closeable closeable, T object) {
 		synchronized (getSynchronizationLock()) {
-			closeableToRef.remove(closeable);
+			return closeableToRef.remove(closeable, object);
```

Comment:
Why is `closeableToRef.remove(closeable)` not enough here? I think it would not matter which `PhantomDelegatingCloseableRef` eventually removes the `closeable` as long as it only gets closed once. The problem I see with the current approach is that it complicates the API (e.g. the dev has to know what `object` actually is).

Reply:
@tillrohrmann It is a paranoid guard against un-register and re-register. After re-registered, there will be a different phantom reference take over this one. Ideally, after unregistered, the phantom reference itself is unreachable from user path, thus that phantom reference should not be enqueued, but I am not that sure about this. If this is not going to happen, `boolean removeCloseableInternal(Closeable closeable)` should be sufficient.

How about adding new tests to test `un-register`/`re-register` and using `closeableToRef.remove(closeable)` ?

Reply:
I think that it should not matter whether the old or new phantom reference gets enqueued and triggers the clean up because the clean up will only be triggered once the object is phantom reachable and there is coming back from this state. Hence, I think we don't need this safeguard if I am not mistaken.

If these methods are not tested yet, then I could help introducing them in order to guard our assumptions here.

Reply:
@tillrohrmann OK. It is just a paranoid concern. I didn't observe this behavior in local test, but if I spy containing registry to keep that phantom reference from unreachable, it indeed close re-registered closeable. This spying behavior will not happen in real code.

I have change the signature to `boolean removeCloseableInternal(Closeable closeable)` in following fixup commit which should be squashed out before merge.

Reply:
@kezhuw For self-review. The paranoid concern does not hold, since java has only one approach to recycle unreachable object, it is overriding `Object.finalize`.